File: src/assets/DSA/LeetCode/0002. Add Two Numbers/addTwoNumbers.java
// import javax.swing.plaf.ListUI;

// class Solution {
//   public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
//     ListNode dummyHead = new ListNode(0);
//     ListNode p = l1, q = l2, curr = dummyHead;
//     int carry = 0;
//     while (p != null || q != null) {
//       int x = (p != null) ? p.val : 0;
//       int y = (q != null) ? q.val : 0;
//       int sum = carry + x + y;
//       carry = sum / 10;
//       curr.next = new ListNode(sum % 10);
//       curr = curr.next;
//       if (p != null)
//         p = p.next;
//       if (q != null)
//         q = q.next;
//     }
//     if (carry > 0) {
//       curr.next = new ListNode(carry);
//     }
//     return dummyHead.next;
//   }
// }
// lets test out the solution
// public class addTwoNumbers {
//   public static void main(String[] args) {
//     Solution solution = new Solution();
//     ListNode l1 = new ListUI(2);
//     l1.next = new ListNode(4);
//     l1.next.next = new ListNode(3);
//     ListNode l2 = new ListNode(5);
//     l2.next = new ListNode(6);
//     l2.next.next = new ListNode(4);
//     ListNode result = solution.addTwoNumbers(l1, l2);
//     while (result != null) {
//       System.out.println(result.val);
//       result = result.next;
//     }
//   }
// }

// Output  7 0 8

// Time Complexity: O(max(m,n)) where m and n represents the length of l1 and l2 respectively.

// Space Complexity: O(max(m,n)) The length of the new list is at most max(m,n)+1.

File: src/assets/5.DSA/LeetCode/0002. Add Two Numbers/addTwoNumbers.js
var addTwoNumbers = function(l1, l2) {

    let array1 = []
    let array2 = []

    while(l1!==null){
        array1.push(l1.val)
        l1 = l1.next
    }

    while(l2!==null){
        array2.push(l2.val)
        l2 = l2.next
    }

    let num1 = array1.reverse().join('')
    let num2 = array2.reverse().join('')


    let resNumber = BigInt(num1) + BigInt(num2)
    let arr3 = String(resNumber).split('').reverse()

    let newHead = new ListNode('')
    let newPointer = newHead

    for(let i=0;i<arr3.length;i++){
        let node = new ListNode(arr3[i])
        newPointer.next=node
        newPointer = node
    }

    return newHead.next

  };


// The current implementation of the function has a time complexity of O(n + m + k), where n is the length of l1, m is the length of l2, and k is the length of the result list. The space complexity is O(n + m + k) as well, due to the use of the two arrays and the result list.

// To optimize for time and space, we can avoid using the two arrays to store the values of l1 and l2, and instead add the corresponding values directly while iterating through the linked lists. This would reduce the space complexity to O(k), where k is the length of the result list, and the time complexity to O(max(n, m) + k), which is better than the original implementation.

// Here's an optimized implementation:

var addTwoNumbers = function(l1, l2) {
    let carry = 0
    let newHead = new ListNode('')
    let newPointer = newHead

    while(l1 || l2 || carry) {
      let val1 = l1 ? l1.val : 0
      let val2 = l2 ? l2.val : 0
      let sum = val1 + val2 + carry

      carry = sum > 9 ? 1 : 0
      let digit = sum % 10

      let newNode = new ListNode(digit)
      newPointer.next = newNode
      newPointer = newNode

      l1 = l1 ? l1.next : null
      l2 = l2 ? l2.next : null
    }

    return newHead.next
  }


//   In this implementation, we use a single while loop to iterate through both l1 and l2 simultaneously, adding the corresponding values and keeping track of the carry. We create the new nodes and add them to the result list as we go. This way, we don't need to store the values of l1 and l2 separately in arrays, and can directly add them to the result list.

